### PR TITLE
Fix `NiconicoIE` login

### PR DESCRIPTION
**Pull request by ekangmonyet**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

niconico now requires MFA when logging in using username and password, breaking the current login method.  This patch updates the login flow and prompts user to input the code.

### CODE REVIEW

File: yt_dlp/extractor/niconico.py [hunk 1]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
-            login_ok = False
-        else:
-            parts = compat_urllib_parse_urlparse(urlh.geturl())
-            if compat_parse_qs(parts.query).get('message', [None])[0] == 'cant_login':
-                login_ok = False
+            }).read().decode('utf-8')
+        need_mfa = 'oneTimePw' in resp_page
+        if need_mfa:
+            mfa_submit = self._search_regex(
+                r'(?<=<form action=["\'])(.*?)(?=["\'])', resp_page, 'mfa_submit',
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:204]:
```suggestion
                r'<form action=["\']([^"\'])["\']', resp_page, 'mfa_submit',
```
why use lookahead/behind?

File: yt_dlp/extractor/niconico.py [hunk 2]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
+                fatal=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:205]: `fatal=True` is default

File: yt_dlp/extractor/niconico.py [hunk 3]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
+            mfa_form_strs = {
+                'otp': self._get_tfa_info('6 digits code')
+            }
+            resp_page = self._request_webpage(
+                'https://account.nicovideo.jp' + mfa_submit, None,
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:210]: fstrings preferred over +

File: yt_dlp/extractor/niconico.py [hunk 4]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
+                note='Performing MFA', errnote='Unable to complete MFA',
+                data=urlencode_postdata(mfa_form_strs),
+                headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                }).read().decode('utf-8')
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:215]: why not download_webpage?

File: yt_dlp/extractor/niconico.py [hunk 5]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
+            need_mfa = 'oneTimePw' in resp_page
+            mfa_error = need_mfa or 'formError' in resp_page
+            if mfa_error:
+                self.report_warning(self._html_search_regex(
+                    '(?<=formError["\']>)(.*?)(?=</div>)', resp_page, 'form_error',
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:220]: same as above

File: yt_dlp/extractor/niconico.py [hunk 6]
```diff
@@ -192,20 +190,38 @@ def _perform_login(self, username, password):
...
+                    default='There\'s an error but the message can\'t be parsed.',
+                    fatal=False, flags=re.DOTALL))
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:222]: default implies non-fatal

**Review by pukkandan — CHANGES_REQUESTED** [review]
cc @animelover1984

**Review by Lesmiscore — APPROVED** [review]
Good work. It worked with my account.

File: yt_dlp/extractor/niconico.py [hunk 7]
```diff
@@ -192,20 +191,39 @@ def _perform_login(self, username, password):
...
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                })
+            need_mfa = 'oneTimePw' in page
+            mfa_error = need_mfa or 'formError' in page
+            if mfa_error:
+                err_msg = self._html_search_regex(
+                    r'formError["\']+>(.*?)</div>', page, 'form_error',
+                    default='There\'s an error but the message can\'t be parsed.',
+                    flags=re.DOTALL)
+                self.report_warning(f'unable to log in: failed mfa, "{err_msg}"')
```
[the remark this answers is not in the document]
Comment (ekangmonyet) [reply yt_dlp/extractor/niconico.py:224]: That seems better, judging by the current usage, I think I will capitalize both `Unable to log in` messages too.

File: yt_dlp/extractor/niconico.py [hunk 8]
```diff
@@ -192,22 +191,41 @@ def _perform_login(self, username, password):
...
+            mfa_form_strs = {
+                'otp': self._get_tfa_info('6 digits code')
+            }
+            page = self._download_webpage(
+                urljoin('https://account.nicovideo.jp', post_url), None,
+                note='Performing MFA', errnote='Unable to complete MFA',
+                data=urlencode_postdata(mfa_form_strs),
+                headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                })
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:216]:
```suggestion
            page = self._download_webpage(
                urljoin('https://account.nicovideo.jp', post_url), None,
                note='Performing MFA', errnote='Unable to complete MFA',
                data=urlencode_postdata({
                    'otp': self._get_tfa_info('6 digits code')
                }), headers={
                    'Content-Type': 'application/x-www-form-urlencoded',
                })
```

File: yt_dlp/extractor/niconico.py [hunk 9]
```diff
@@ -192,22 +191,41 @@ def _perform_login(self, username, password):
...
-            login_ok = False
-        else:
-            parts = compat_urllib_parse_urlparse(urlh.geturl())
-            if compat_parse_qs(parts.query).get('message', [None])[0] == 'cant_login':
-                login_ok = False
+        need_mfa = 'oneTimePw' in page
+        if need_mfa:
+            post_url = self._search_regex(
+                r'<form[^>]+action=(["\'])(?P<url>.+?)\1', page,
+                'post url', group='url')
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:206]:
```suggestion
        if 'oneTimePw' in page:
            post_url = self._search_regex(
                r'<form[^>]+action=(["\'])(?P<url>.+?)\1', page, 'post url', group='url')
```

File: yt_dlp/extractor/niconico.py [hunk 10]
```diff
@@ -192,22 +191,41 @@ def _perform_login(self, username, password):
...
+            page = self._download_webpage(
+                urljoin('https://account.nicovideo.jp', post_url), None,
+                note='Performing MFA', errnote='Unable to complete MFA',
+                data=urlencode_postdata(mfa_form_strs),
+                headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                })
+            need_mfa = 'oneTimePw' in page
+            mfa_error = need_mfa or 'formError' in page
+            if mfa_error:
```
Comment (pukkandan) [inline yt_dlp/extractor/niconico.py:219]:
```suggestion
            if 'oneTimePw' in page or 'formError' in page:
```

**Pull request merged**